### ISSUE TITLE
Trigger something as soon as markdown has been fully parsed

### ISSUE THREAD

**Issue by SoftCreatR** [showdownjs/showdown#556]
I am currently using showdown in combination with ng-showdown and i try to modify code boxes to be compatible with [prism.js](https://prismjs.com) / [ng-prism](https://github.com/joswhite/ng-prism).

For this, i've created an extension that adds the class `code` to code tags. This seems to work (even if this might not be the best implementation, yet). Additionally, i've created an extension as listener, that is supposed to run `Prism.highlightAll()` whenever code blocks have been modified by showdown. However, it doesn't seem to work that way. But if i execute `Prism.highlightAll()` in the browser console after loading the page, it works, so the first extension does it's job.

This is my current code:

```
/* eslint-disable */
// jshint ignore: start
(function(){
    'use strict';

    angular.module('libraryView')
        .constant('showdownPrismPrepare',
            function showdownPrism() {
                return [
                    {
                        type: 'output',
                        filter: function (text) {
                            // use new shodown's regexp engine to conditionally parse codeblocks
                            var left  = '<pre><code\\b[^>]*>',
                                right = '</code></pre>',
                                flags = 'g',
                                replacement = function (wholeMatch, match, left, right) {
                                    // unescape match to prevent double escaping
                                    match = match.replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>');

                                    left = left.replace('class="', 'class="code ');

                                    return left + match + right;
                                };

                            return showdown.helper.replaceRecursiveRegExp(text, replacement, left, right, flags);
                        }
                    }
                ];
            }
        )
        .constant('showdownPrismHighlight',
            function showdownPrismHighlight() {
                return [
                    {
                        type: 'listener',
                        listeners: {
                            'encodeCode.after': function (event, text, options, globals) {
                                Prism.highlightAll();
                            }
                        }
                    }
                ]
            }
        );
})();

// ...

$showdownProvider.loadExtension(showdownPrismPrepare);
$showdownProvider.loadExtension(showdownPrismHighlight);
```

Due to the fact, that this isn't an ng-showndown specific problem, i've created the issue here. I hope, that's okay.

So the question is: What am i doing wrong? Or should i switch back to hljs (because there is at least one working extension). Or is there a better approach to

1. add the class `code` to generated code tags
2. execute the function (once, if possible)

Thank you for your time.

**Comment by tivie** [comment]
# TL;DR;

Although your event order is wrong, and it's perfectly possible to call Prism from within Showdown, it won't achieve the effect you're after.

The correct order of things that must happen is:

 - [X] Feed the input into showdown
 - [X] Perform the desired alterations with an extension
 - [ ] Get the output from showdown
 - [ ] Insert the output into the page
 - [ ] Run Prism

So, as you see, a showdown extension will never be able to solve your issue, since showdown has to finish execution AND you need to insert it's output into the page in order for Prism to do its thing.

HOWEVER, nothing prevents you from using an angular event to run prism, AFTER showdown performs it's modifications.

-----

There are a few problems here, I will try to walk you through them:

## Event order

The event `encodeCode.after` is triggered after showdown's `encodeCode` subparser is run.
The output extension is run right at the end of everything.
This means `encodeCode.after` happens BEFORE output extensions.

## How showdown  work

Showdown takes a piece of text (in markdown) and converts it to HTML. 
When you can the `converter.makeHtml()` it **returns the converted text**. Showdown doesn't insert anything in the page.
So, basically, until you insert it in the page, prism won't see it and will not perform any alterations on the code.
I'm not sure if it's possible to feed HTML to Prism and get the output, though.


## Showdown supports prism out of the box... with a caveat

As you can see in this fiddle, showdown supports prism out of the box. Without you requiring to perform any alterations to showdown's output.

https://jsfiddle.net/tivie/fphe4vsn/

At least for GH Style blocks where you specify a language. Indented code blocks, since they don't have language attribute, prism won't work out of the box.

**Comment by tivie** [comment]
Closing as no update for a while. Feel free to reopen if you think you need more help.

---

> ### Donate [![Click here to lend your support to: ShowdownJS website and testing platform and make a donation at paypal.com](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/tiviesantos)
>As you know, ShowdownJS is a free library and it will remain free forever. However, maintaining and improving the library costs time and money.
>If you like our work and find our library useful, please donate ~~[through Pledgie](https://pledgie.com/campaigns/35166) or~~ directly [through paypal](https://www.paypal.me/tiviesantos)!! Your contribution will be greatly appreciated and help us continue to develop this awesome library.
